### ISSUE TITLE
Performance issue with min and max methods in dart:math 

### ISSUE THREAD

**Issue by julemand101** [dart-lang/sdk#34090]
**Dart SDK Version (`dart --version`)**
_Problematic version:_
Dart VM version: 2.0.0 (Fri Aug 3 10:53:23 2018 +0200) on "windows_x64"

_Compared against:_
Dart VM version: 1.24.3 (Wed Dec 13 18:05:29 2017) on "windows_x64"

**Whether you are using Windows, MacOSX, or Linux (if applicable)**
Windows 10 64 bit.

**Whether you are using Chrome, Safari, Firefox, Edge (if applicable)**
Running as standalone with dart.exe

----

When running the following code in Dart 2.0.0 and Dart 1.24.3:
```dart
import 'dart:math' as math;
main() {
  math.Random rand = new math.Random(1);
  List<int> list = new List.generate(100000000, (_) => rand.nextInt(100000000));
  var t1 = new DateTime.now();
  print(list.reduce(math.max));
  var t2 = new DateTime.now();
  print("Time: ${t2.difference(t1)}");
}
```
I got the following timings:

Dart 1.24.3 | Dart 2.0.0
------------ | -------------
0:00:00.550652 | 0:00:24.152452

With the observatory I can see the CPU usage in Dart 2.0 are used on something called DRT_SubtypeCheck:

![image](https://user-images.githubusercontent.com/116227/43801644-fbe68e12-9a93-11e8-8af2-04bfad8e81ee.png)

If I modify the code to not use the math.max method:
```dart
import 'dart:math' as math;
main() {
  math.Random rand = new math.Random(1);
  List<int> list = new List.generate(100000000, (_) => rand.nextInt(100000000));
  var t1 = new DateTime.now();
  print(list.reduce((a, b) => a > b ? a : b));
  var t2 = new DateTime.now();
  print("Time: ${t2.difference(t1)}");
}
```
Then I gets the following timings:

Dart 1.24.3 | Dart 2.0.0
------------ | -------------
0:00:00.530667 | 0:00:00.536129

I should note that the same issue can also be reproduced for math.min.

**Comment by mraleph** [comment]
We are working on this. We need to introduce checked and unchecked entries into closures for this to become faster.

**Comment by sjindel-google** [comment]
The remaining regression appears to be due to the fact that we aren't optimizing the arithmetic inside `max` as well as in the local closure, since the parameter types are only known to be `num`.

**Comment by mraleph** [comment]
@sjindel-google I think it is actually [this chunk of code](https://github.com/dart-lang/sdk/blob/master/runtime/vm/compiler/call_specializer.cc#L1584-L1617) - have special optimization for `min`/`max` which only triggers if no type arguments are passed. 

We should figure out if we can make that work when type arguments are passed as well.

**Comment by aartbik** [comment]
For JIT, I prototyped mraleph's idea by inlining MathMinMax even when the type parameter is present. This yields the following intermediate.
```
B32[target]:154
 AssertSubtype:148(T, num, 'T', instantiator_type_args(v0), function_type_args(v41))
 goto:160 B34
B33[target]:156
 goto:162 B34
B34[join]:158 pred(B32, B33)
 CheckSmi:164(v3)
 AssertAssignable:164(v3 T{_Smi!}, T, 'a', instantiator_type_args(v0), function_type_args(v41)) T{_Smi!}
 CheckSmi:166(v4)
 AssertAssignable:166(v4 T{_Smi!}, T, 'b', instantiator_type_args(v0), function_type_args(v41)) T{_Smi!}
 v45 <- MathMinMax:168(v3 T{_Smi!}, v4 T{_Smi!}) [-4611686018427387904, 4611686018427387903] T{_Smi!}
```
This increased performance a bit, but not as much as I had hoped. Profiling the generated JIT code reveals that we spent more time in the AssertAssignable code than in the Max code itself. So next we should explore if this can be further simplified.

**Comment by julemand101** [comment]
Well, from my point of view the issue seems to be fixed in Dart 2.1 and I am closing this issue. You are welcome to reopen the issue if you guys thinks there are still more there can be done on this matter.
